router/src/server.rs

```diff
@@ -114,6 +115,150 @@ async fn get_model_info(info: Extension<Info>) -> Json<Info> {
     Json(info.0)
 }
 
+#[utoipa::path(
+    post,
+    tag = "Text Generation Inference",
+    path = "/chat_tokenize",
+    request_body = ChatRequest,
+    responses((status = 200, description = "Templated and tokenized ChatRequest", body = ChatTokenizeResponse))
+)]
+async fn get_chat_tokenize(
+    Extension(infer): Extension<Infer>,
+    Json(req): Json<ChatRequest>,
+) -> Result<(HeaderMap, Json<ChatTokenizeResponse>), (StatusCode, Json<ErrorResponse>)> {
+    metrics::counter!("tgi_request_count").increment(1);
+
+    let ChatRequest {
+        model,
+        max_tokens,
+        messages,
+        seed,
+        stop,
+        stream,
+        tools,
+        tool_choice,
+        tool_prompt,
+        temperature,
+        response_format,
+        ..
+    } = req;
+
+    if response_format.is_some() && tools.is_some() {
+        return Err((
+            StatusCode::UNPROCESSABLE_ENTITY,
+            Json(ErrorResponse {
+                error: "Grammar and tools are mutually exclusive".to_string(),
+                error_type: "validation".to_string(),
+            }),
+        ));
+    }
+
+    let tool_grammar = match ToolGrammar::apply(tools, tool_choice) {
+        Ok(grammar) => grammar,
+        Err(err) => {
+            metrics::counter!("tgi_request_failure", "err" => "validation").increment(1);
+            tracing::error!("{}", err);
+            return Err((
+                StatusCode::UNPROCESSABLE_ENTITY,
+                Json(ErrorResponse {
+                    error: err.to_string(),
+                    error_type: err.error_type().to_string(),
+                }),
+            ));
+        }
+    };
+
+    let tools_grammar_prompt = tool_grammar.as_ref().map(|t| {
+        (
+            GrammarType::Json(serde_json::json!(t)),
+            tool_prompt.unwrap_or_default(),
+        )
+    });
+
+    let (tools_grammar_prompt, _grammar) = response_format
+        .map(|rf| (None, Some(rf)))
+        .unwrap_or_else(|| {
+            (
+                tools_grammar_prompt.clone(),
+                tools_grammar_prompt.map(|(g, _)| g),
+            )
+        });
+
+    let inputs = match infer.apply_chat_template(messages, tools_grammar_prompt) {
+        Ok(inputs) => inputs,
+        Err(err) => {
+            metrics::counter!("tgi_request_failure", "err" => "validation").increment(1);
+            tracing::error!("{}", err);
+            return Err((
+                StatusCode::UNPROCESSABLE_ENTITY,
+                Json(ErrorResponse {
+                    error: err.to_string(),
+                    error_type: err.error_type().to_string(),
+                }),
+            ));
+        }
+    };
```

Comment:
This code is already used in some validation, we should probably extract it into a function (so this doesn't drift from the original implem later).

Also this code is very hard to read for me (lots of options everywhere). I'm sure we can simplify it a lot.

Reply:
totally agree! in the latest commit i've moved this code into a function called `prepare_chat_input` which is reusable and much easier to read.